src/internal/operators/timeout.ts

```diff
@@ -332,62 +325,71 @@ export function timeout<T, R, M>(
 
     return lift(source, function (this: Subscriber<T | R>, source: Observable<T>) {
       const subscriber = this;
-      const subscription = new Subscription();
-      let innerSub: Subscription;
-      let timerSubscription: Subscription | null = null;
+      // This subscription encapsulates our subscription to the
+      // source for this operator. We're capturing it separately,
+      // because if there is a `with` observable to fail over to,
+      // we want to unsubscribe from our original subscription, and
+      // hand of the subscription to that one.
+      let originalSourceSubscription: Subscription;
+      // The subscription for our timeout timer. This changes
+      // every time get get a new value.
+      let timerSubscription: Subscription;
+      // A bit of state we pass to our with and error factories to
+      // tell what the last value we saw was.
       let lastValue: T | null = null;
+      // A bit of state we pass to the with and error factories to
+      // tell how many values we have seen so far.
       let seen = 0;
       const startTimer = (delay: number) => {
-        subscription.add(
+        subscriber.add(
           (timerSubscription = scheduler!.schedule(() => {
             let withObservable: Observable<R>;
-            const info: TimeoutInfo<T, M> = {
-              meta,
-              lastValue,
-              seen,
-            };
             try {
-              withObservable = from(_with!(info));
+              withObservable = from(
+                _with!({
+                  meta,
+                  lastValue,
+                  seen,
+                })
+              );
             } catch (err) {
               subscriber.error(err);
               return;
             }
-            innerSub.unsubscribe();
-            subscription.add(withObservable.subscribe(subscriber));
+            originalSourceSubscription.unsubscribe();
+            withObservable.subscribe(subscriber);
           }, delay))
         );
       };
 
-      subscription.add(
-        (innerSub = source.subscribe({
-          next: (value) => {
-            timerSubscription?.unsubscribe();
-            timerSubscription = null;
-            seen++;
-            lastValue = value;
-            if (each != null && each > 0) {
-              startTimer(each);
+      subscriber.add(
+        (originalSourceSubscription = source.subscribe(
+          new OperatorSubscriber(
+            subscriber,
+            (value) => {
+              // clear the timer so we can emit and start another one.
+              timerSubscription.unsubscribe();
+              seen++;
+              // Emit
+              subscriber.next((lastValue = value));
+              each && each > 0 && startTimer(each);
```

Comment:
nitpick: neither `null` nor `undefined` are `> 0`, so the initial `each &&` could be removed.